File: impl/src/middleware/action/fromPersonAction.ts
import {MutationMiddleware, MutationResult} from "../_type";
import {JsonUtil} from "../../utils/util";
import {JsonSchema} from "../../utils/JsonSchema";
import {PersonQueryType} from "../../utils/person/setProperty";

export type FromPersonQueryElement = JsonSchema["children"]["rule_group"]["children"]["action_rule"]["children"]["from_person"]
export type OnPersonRuleQueryElement = FromPersonQueryElement["children"]["on_person"]
export type FromPersonSameLocationGraphNode = OnPersonRuleQueryElement["children"]["selection"]["children"]["from_person_same_location_graph_node"];

const onPersonHandle = (writeUnit: JsonUtil, onPersonQueryElement: OnPersonRuleQueryElement, originPerson: PersonQueryType, targetPerson: PersonQueryType): MutationResult[] => {
  return onPersonQueryElement
    ?.queryOptional("mutations")
    ?.queryAllOptional("property_mutation")
    ?.flatMap(property_mutation => {
      return writeUnit.person.applyPropertyMutation(targetPerson, property_mutation, targetPerson, originPerson);
    });
}


const isFromPersonSameLocationGraphNode = (writeUnit: JsonUtil, originPerson: PersonQueryType, targetPerson: PersonQueryType, fromPersonSameLocationGraphNode?: FromPersonSameLocationGraphNode): boolean => {
  if (!fromPersonSameLocationGraphNode) {
    return true;
  }
  const originLocation = writeUnit.locationGraph.findPersonLocation(originPerson.attributeMap.id);
  const targetLocation = writeUnit.locationGraph.findPersonLocation(targetPerson.attributeMap.id);

  if (originLocation.length === 0 || targetLocation.length === 0) {
    return false;
  }

  const commonLocation = originLocation.find(originLocation => {
    return targetLocation.find(targetLocation => {
      return originLocation?.locationGraph?.attributeMap?.id === targetLocation?.locationGraph?.attributeMap?.id
        && originLocation?.node?.attributeMap?.id === targetLocation?.node?.attributeMap?.id
    })
  })
  const applicable = fromPersonSameLocationGraphNode?.attributeMap?.value === "true";
  if (applicable) {
    return !!commonLocation;
  } else {
    return !commonLocation;
  }
}

const onPersonApplicable = (writeUnit: JsonUtil, originPerson: PersonQueryType, targetPerson: PersonQueryType, onPersonQueryElement: OnPersonRuleQueryElement,): boolean => {
  const selection = onPersonQueryElement.queryOptional("selection");
  const fromPersonSameLocationGraphNode = selection?.queryOptional("from_person_same_location_graph_node")
  if (!isFromPersonSameLocationGraphNode(writeUnit, originPerson, targetPerson, fromPersonSameLocationGraphNode)) {
    return false;
  }
  if (!writeUnit.person.isSelectionApplicableTo(selection, targetPerson)) {
    return false;
  }
  return true;

}


export const fromPersonActionMiddleware: MutationMiddleware = readJson => {
  const actionMetadataList = readJson.getRuleGroups()
    .flatMap(rule_group => rule_group.queryAllOptional("action_rule"))
    .flatMap(action_rule => action_rule.queryAllOptional("from_person"));

  const actionElementList = readJson.json.queryOptional("actions")
    ?.queryAllOptional("from_person")

  return async writeUnit => {
    if (!(actionElementList)) {
      return;
    }

    const people = writeUnit.json.queryOptional("data")?.queryOptional("people");
    if (!people) {
      return
    }


    const result = actionElementList.flatMap(actionElement => {
      const actionRule = actionMetadataList.find(metadata => metadata.attributeMap.id === actionElement.attributeMap.from_person_rule_ref);
      if (!actionRule) {
        return [];
      }
      const action: Array<(writeUnit: JsonUtil) => Promise<void>> = [];
      const personList = people.queryAllOptional("person");
      const selfPerson = personList.find(person => person.attributeMap.id === actionElement.attributeMap.person_id_ref);

      const selfSelection = actionRule.queryOptional("selection");
      if (selfSelection) {
        if (!writeUnit.person.isSelectionApplicableTo(selfSelection, selfPerson)) {
          return
        }
      }

      const result = actionElement.queryAllOptional("on_person")
        .flatMap(on_person => {
          const onPersonRule = actionRule.queryOptional("on_person");
          const targetPerson = personList.find(person => person.attributeMap.id === on_person.attributeMap.person_id_ref);

          if (!onPersonApplicable(writeUnit, selfPerson, targetPerson, onPersonRule)) {
            return;
          }

          let selfResult = actionRule.queryOptional("mutations")
            ?.queryAllOptional("property_mutation")
            ?.map(property_mutation => {
              return writeUnit.person.applyPropertyMutation(selfPerson, property_mutation, selfPerson, targetPerson);
            })

          const handleResult = onPersonHandle(writeUnit, onPersonRule, selfPerson, targetPerson);
          return [...(selfResult ?? []), ...(handleResult ?? [])]
        })
      action.push(...result);
      return action;
    });

    await Promise.all(result.map(e => e?.(writeUnit)));

    writeUnit.json.queryOptional("actions")
      ?.queryAllOptional("from_person")
      ?.forEach(from_person => from_person.removeFromParent());
  }
}

File: implementation/src/main/java/ro/anud/xml_xsd/implementation/middleware/action/FromPersonAction.java
package ro.anud.xml_xsd.implementation.middleware.action;

import org.springframework.stereotype.Service;
import ro.anud.xml_xsd.implementation.model.WorldStep.Actions.Actions;
import ro.anud.xml_xsd.implementation.model.WorldStep.Actions.FromPerson.FromPerson;
import ro.anud.xml_xsd.implementation.model.WorldStep.Data.People.Person.Person;
import ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.ActionRule.FromPerson.Mutations.Mutations;
import ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.ActionRule.FromPerson.OnPerson.OnPerson;
import ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.ActionRule.FromPerson.OnPerson.Selection.FromPersonSameLocationGraphNode.FromPersonSameLocationGraphNode;
import ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.ActionRule.FromPerson.OnPerson.Selection.Selection;
import ro.anud.xml_xsd.implementation.model.WorldStep.WorldStep;
import ro.anud.xml_xsd.implementation.service.Mutation;
import ro.anud.xml_xsd.implementation.service.WorldStepInstance;

import java.util.stream.Stream;

import static ro.anud.xml_xsd.implementation.util.logging.LogScope.logScope;

@Service
public class FromPersonAction {

    public static void apply(final WorldStepInstance worldStepInstance) {
        try (var scope = logScope()) {
            var ruleRepository = worldStepInstance.ruleRepository;
            var personRepository = worldStepInstance.person.repository;

            var fromPersonStream = worldStepInstance.streamWorldStep()
                    .flatMap(WorldStep::streamActions)
                    .flatMap(Actions::streamFromPerson)
                    .toList();
            if (fromPersonStream.isEmpty()) {
                scope.log("fromPerson list is empty");
                return;
            }


            worldStepInstance.getOutInstance()
                    .streamWorldStep()
                    .flatMap(WorldStep::streamActions)
                    .flatMap(Actions::streamFromPerson)
                    .toList()
                    .forEach(FromPerson::removeFromParent);


            fromPersonStream.forEach(actionElement -> {
                try (var localLogger = logScope(actionElement.buildPath())){
                    localLogger.log("fromPersonRuleRef", actionElement.getFromPersonRuleRef());
                    localLogger.log("personIdRef", actionElement.getPersonIdRef());
                    var actionRuleOptional = ruleRepository.getPersonById(actionElement.getFromPersonRuleRef());
                    if (actionRuleOptional.isEmpty()) {
                        localLogger.log("early return actionRule is empty");
                        return;
                    }
                    var actionRule = actionRuleOptional.get();
                    var selfPersonOptional = personRepository.personById(actionElement.getPersonIdRef());
                    if (selfPersonOptional.isEmpty()) {
                        localLogger.log("early return actionRule is empty");
                        return;
                    }
                    var selfPerson = selfPersonOptional.get();
                    var selfSelection = actionRule.getSelection();
                    if (selfSelection.isPresent()) {
                        var result = worldStepInstance.person.selection.isSelectionApplicableTo(
                                selfSelection.get(),
                                selfPersonOptional.get()
                        );

                        if (!result) {
                            localLogger.log("early return isSelectionNotApplicable is ", result);
                            return;
                        }
                    }
                    actionElement.streamOnPerson()
                            .forEach(onPerson -> {
                                try(var localLocalLogger = logScope()) {
                                    var onPersonRule = actionRule.getOnPerson();
                                    var targetPerson = personRepository.personById(onPerson.getPersonIdRef());
                                    if (onPersonRule.isEmpty() || targetPerson.isEmpty()) {
                                        localLocalLogger.log(
                                                "onPersonRule:", onPersonRule,
                                                "targetPerson", targetPerson
                                        );
                                        return;
                                    }
                                    if (!onPersonApplicable(
                                            worldStepInstance,
                                            selfPerson,
                                            targetPerson.get(),
                                            onPersonRule.get()
                                    )) {
                                        localLocalLogger.log("onPersonApplicable false");
                                        return;
                                    }
                                    Stream<Mutation<Person>> fromMutationResult = actionRule.streamMutations()
                                            .flatMap(Mutations::streamPropertyMutation)
                                            .map(typePropertyMutation -> worldStepInstance.person
                                                    .applyPropertyMutation(
                                                            selfPerson,
                                                            typePropertyMutation,
                                                            selfPerson,
                                                            targetPerson.get()
                                                    )
                                            );
                                    Stream<Mutation<Person>> onPersonMutationResult = onPersonRule.flatMap(OnPerson::getMutations)
                                            .stream()
                                            .flatMap(ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.ActionRule.FromPerson.OnPerson.Mutations.Mutations::streamPropertyMutation)
                                            .map(typePropertyMutation -> worldStepInstance
                                                    .person
                                                    .applyPropertyMutation(
                                                            targetPerson.get(),
                                                            typePropertyMutation,
                                                            targetPerson.get(),
                                                            selfPerson
                                                    )
                                            );

                                    Stream.concat(fromMutationResult, onPersonMutationResult)
                                            .forEach(mutation -> {
                                                mutation.apply(worldStepInstance.getOutInstance());
                                            });
                                }

                            });
                }

            });
        }

    }

    private static boolean onPersonApplicable(
        final WorldStepInstance worldStepInstance,
        final Person originPerson,
        final Person targetPerson,
        final OnPerson onPersonRule) {
        try (var scope = logScope("originPerson", originPerson.getId(), "targetPerson", targetPerson.getId())){
            var selection = onPersonRule.getSelection();
            if (selection.isEmpty()) {
                return true;
            }
            var isNotFromSameLocation = selection.flatMap(Selection::getFromPersonSameLocationGraphNode)
                    .map(fromPersonSameLocationGraphNode -> !isFromPersonSameLocationGraphNode(
                            worldStepInstance,
                            originPerson,
                            targetPerson,
                            fromPersonSameLocationGraphNode)
                    )
                    .orElse(false);

            if (isNotFromSameLocation) {
                scope.log("isNotFromSameLocation");
                return scope.logReturn(false);

            }
            var selectionInstance = worldStepInstance.person.selection;
            var invalidSelectionList = selection
                    .stream()
                    .filter(selection1 -> !selectionInstance.isSelectionApplicableTo(
                            selection1,
                            targetPerson)
                    )
                    .findAny();

            if (invalidSelectionList.isPresent()) {
                scope.log("invalid selection found");
                return scope.logReturn(false);
            }
            return scope.logReturn(true);
        }

    }

    private static boolean isFromPersonSameLocationGraphNode(
        final WorldStepInstance worldStepInstance,
        final Person selfPerson,
        final Person targetPerson,
        final FromPersonSameLocationGraphNode fromPersonSameLocationGraphNode) {
        try (var scope = logScope()){
            var locationGraphInstance = worldStepInstance.locationGraph;
            var originLocation = locationGraphInstance.findPersonLocation(selfPerson.getId());
            var targetLocation = locationGraphInstance.findPersonLocation(targetPerson.getId());

            if (originLocation.isEmpty() || targetLocation.isEmpty()) {
                scope.log("originLocation or targetLocation is empty");
                return scope.logReturn(false);
            }

            var commonLocation = originLocation.stream().anyMatch(originFindPersonResult -> targetLocation.stream()
                    .anyMatch(targetFindPersonResult -> {
                        var originNode = originFindPersonResult.node();
                        var targetNode = originFindPersonResult.node();
                        if (originNode.isEmpty() || targetNode.isEmpty()) {
                            return false;
                        }
                        return originFindPersonResult.locationGraph().getId() == targetFindPersonResult.locationGraph().getId()
                                && originNode.get().getId().equals(targetNode.get().getId());

                    })
            );
            var applicable = fromPersonSameLocationGraphNode.getValue().equals("true");
            if (applicable) {
                return scope.logReturn(commonLocation);
            }
            return scope.logReturn(!commonLocation);
        }

    }
}
